Code action to add/delete imports

youben11 opened the pull request:
This is based on the previous work of @gatesn 

- Actions to import suggested modules based on the index of symbols of all available python modules (might be slow to build but cached for future use)
- Linting hint about unresolved symbols only as we aren't sure if those symbols aren't just variables.

#### Last features

- Actions to remove unreferenced imports
- Linting warning about unreferenced imports, variables and functions

related #86

youben11 committed:
Auto import actions using importmagic

youben11 committed:
import re

youben11 committed:
typo

youben11 committed:
test against a specific action

youben11 committed:
disable actions test

youben11 committed:
missing white-space

youben11 committed:
separate action generation logic: fix linting problem

youben11 committed:
fix: pass document instead of source

youben11 committed:
filter results based on minScore parameter

youben11 committed:
update test

youben11 commented:
@gatesn is there any way to make the call to pyls_code_actions non blocking?

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,187 @@
+        log.debug("Started building importmagic index")
+        index = importmagic.SymbolIndex()
+        # The build tend to be noisy
+        logging.getLogger('importmagic.index').setLevel(logging.ERROR)
```
Think we should leave log levels up to the operator

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,187 @@
+    """
+    key = tuple(sys_path)
+    if key not in _index_cache:
+        log.debug("Started building importmagic index")
```
Maybe an info level log?

youben11 replied:
Was using them for seeing if the call is blocking and calculating the time of the index build, but yeah, an INFO level log would make more sense.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,187 @@
+        A list of dictionaries.
+    """
+    scope = importmagic.Scope.from_source(document.source)
+    unresolved, _unreferenced = scope.find_unresolved_and_unreferenced_symbols()
```
Maybe file a follow-up ticket to support a code action for removing unreferenced imports?

youben11 replied:
Nice feature to have, will work on it.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,187 @@
+    return actions
+
+
+def generate_add_action(document, index, module, variable):
```
Prefix with `_` for a "private" function

gatesn committed:
Merge branch 'develop' into importmagic

youben11 committed:
private function

youben11 committed:
change log level

youben11 committed:
Merge branch 'importmagic' of github.com:youben11/python-language-server into importmagic

zewa666 commented:
@youben11 I've checked out your version and tried it with a Webdriver/Selenium sample. Works fantastic to e.g autoimport `from selenium import webdriver`. What doesn't work though is importing local references from another file. Is that the supposed way how importmagic works?

youben11 committed:
warn about unreferenced import/variable/function

youben11 commented:
Hi @zewa666, thank you for testing this functionality and providing feedback.

Yeah the index of symbols is built from the python modules found under sys.path, but I will certainly add the local workspace so you can have suggestions to import from project files.

youben11 committed:
remove unreferenced imports

gatesn commented on pyls/_utils.py:
```diff
@@ -198,3 +199,7 @@ def is_process_alive(pid):
+
+
+def camel_to_underscore(camelcase):
+    return re.sub('([A-Z]+)', r'_\1', camelcase).lower()
```
I think this would rewrite `HelloWorld` to `_hello_world`. Can we be more specific, e.g. this: https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+_index_cache = {}
+
+
+def _get_index(sys_path):
```
Let's make this return a future, that way we can skip results instead of blocking on indexing.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+                    'end': {'line': line_no, 'character': pos + len(unres)}
+                },
+                'message': "Unresolved import '%s'" % unres,
+                'severity': lsp.DiagnosticSeverity.Hint,
```
Should this not be a warning / error? Or does it overlap with e.g. pyflakes results

youben11 replied:
This is a hint because we actually don't know if this is a module to be imported or just a non defined variable, so better not to trigger a false positive. The unreferenced symbols are flagged with Warning because we know in advance if the symbol is either an import or a varialbe/function and the error should be consistent.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+        if diagnostic.get('source') != SOURCE:
+            continue
+        message = diagnostic.get('message', '')
+        if message.startswith('Unreferenced'):
```
Why are we doing a `startswith` _and_ a regex match? Isn't one sufficient?

youben11 replied:
True, should be cleaner now.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+    index = _get_index(sys.path)
+    actions = []
+    diagnostics = context.get('diagnostics', [])
+    for diagnostic in diagnostics:
```
I wonder if it's easier to just call `pyls_lint(document)`, then try to find the context diagnostic in the results. That way we don't have to parse the messages, just have to do an equality check.

e.g. `action_diags = [diag for diag in pyls_lint(document) if diag in context.get('diagnostics', [])]`

youben11 replied:
True that the diagnostics list can grow bigger than the local diagnostics and we will just pass most of them. Also we don't need to do the equality check if we call pyls_lint() as all the diagnostics will be processed.

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+            'newText': args['newText']
+        }]
+    }]}
+    workspace.apply_edit(edit)
```
I think newer versions of clients support putting the edit in the code action:

```
/**
 * A code action represents a change that can be performed in code, e.g. to fix a problem or
 * to refactor code.
 *
 * A CodeAction must set either `edit` and/or a `command`. If both are supplied, the `edit` is applied first, then the `command` is executed.
 */
export interface CodeAction {

	/**
	 * A short, human-readable, title for this code action.
	 */
	title: string;

	/**
	 * The kind of the code action.
	 *
	 * Used to filter code actions.
	 */
	kind?: CodeActionKind;

	/**
	 * The diagnostics that this code action resolves.
	 */
	diagnostics?: Diagnostic[];

	/**
	 * The workspace edit this code action performs.
	 */
	edit?: WorkspaceEdit;

	/**
	 * A command this code action executes. If a code action
	 * provides an edit and a command, first the edit is
	 * executed and then the command.
	 */
	command?: Command;
}
```

But I'm not sure which client capability we should check. What you've got now is probably safer

gatesn commented on pyls/plugins/importmagic_lint.py:
```diff
@@ -0,0 +1,264 @@
+
+
+def _remove_command_title(import_name):
+    return 'Remove import of "%s"' % import_name
```
Maybe `Remove unused import of %s`

gatesn commented on test/test_utils.py:
```diff
@@ -87,3 +87,9 @@ def test_clip_column():
+def test_camel_to_underscore():
+    assert _utils.camel_to_underscore('camelCase') == 'camel_case'
+    assert _utils.camel_to_underscore('hangClosing') == 'hang_closing'
+    assert _utils.camel_to_underscore('ignore') == 'ignore'
```
Can you test a few more cases, e.g. empty string, `UpperCamel`, etc.

youben11 replied:
I thought camelCase always starts with a lowercase but looks like their is SomeVariants.

gatesn commented on vscode-client/package.json:
```diff
@@ -35,6 +35,11 @@
                     },
                     "uniqueItems": true
                 },
+                "pyls.plugins.importmagic_lint.minScore": {
```
Can you add the `pyls.plugins.importmagic_lint.enabled` option

youben11 committed:
config option to enable importmagic

youben11 committed:
fix camel_to_underscore

youben11 committed:
clarify remove command action title

youben11 committed:
clean message checking

youben11 committed:
get diagnostics from pyls_lint

youben11 committed:
non blocking index build

youben11 committed:
store cache as dict

youben11 committed:
test remove action

youben11 committed:
better linting test

youben11 commented:
The importmagic doesn't return the location of unreferenced/unresolved symbols so we needed to look them up ourselves, the search logic was a simple naive .find() call which might return false positives (e.g 'os' is found in 'pos'). Next I will be looking on how to simply tokenize the source code and search among the tokens so we make sure the matching is correct.

youben11 committed:
search symbols in source tokens instead of raw string:

raw string matching lead to false positives (e.g 'os' if found under
'pos' but it's not the actual symbol that we were searching).

youben11 committed:
check for None value

youben11 committed:
fix range

gatesn commented:
@youben11 perhaps you could use Jedi's `names()` call instead of walking the AST yourself. It may help, haven't verified: https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.names

youben11 commented:
@gatesn the disadvantage with jedi.names is that it does only return definitions, symbols such as function calls aren't returned. The implementation using tokenize is pretty straightforward, however, I have to deal with the two different version of python here as the function takes different parameters for py2 and py3.

#### Update
Actually there is a backward compatible function to use that works for both version.

youben11 committed:
use backward compatible tokenizer func

youben11 committed:
fix indexing and pattern matching

youben11 committed:
typo

youben11 committed:
new style class

youben11 committed:
use list comprehension for better perf

gjabell commented:
Any motion on this PR? Would love to have this functionality in the language server!

weeman1337 commented:
Hi there! Thanks for working on the import function :+1: 

After reading the comments it is not clear to me what is left to do. Since I really want this feature I would be happy to contribute - if I would know hat to do :D

ccordoba12 commented:
This project is not maintained anymore. However, the community created a fork of it at

https://github.com/python-lsp/python-lsp-server/

So please submit this PR there if you want to see it finally merged.

Mte90 commented:
If someone is looking for the same feature in the latest pylsp: https://github.com/python-lsp/python-lsp-server/pull/199
